compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -215,9 +260,65 @@ bool TaggedShapeAnalyzer::verify_tag() const
   return true;
 }
 
+/**
+ * @brief Initialize the class members and check under conditions
+ *
+ *  Condtiions that have to be met for analyzer
+ *    c1: input rank >= output rank
+ *    c2: The 'perm' of tranpose should be a CircleConst* type
+ *    c3: The shapes of the given nodes should be all known
+ *
+ * @return True, if all conditions are satisfied and class members are initialized successfully
+ *         False, otherwise
+ */
+template <loco::DataType DType>
+bool TaggedShapeAnalyzer::init(const luci::CircleTranspose *front_transpose,
+                               const luci::CircleReshape *mid_reshape,
+                               const luci::CircleTranspose *back_transpose)
```

Comment:
(note) 
This is the main change - introduce this `init()` function to explicitly check the analyzer's working condition. 

* Before 
   * conditions(`c1` and `c2`)  are checked in callers and some conditions(`c3`) are not clearly checked 
  
* After
   * all conditions are checked in this `init()` function. 
   * if `init()` function fails it means that - the analyzer didn't support this case

Reply:
This change looks good to me. It's much clearer than before.

I think `DType` is not a good template parameter. It's the dtype of indices of reshape and dtype of indices of transpose. They don't have to be the same, so the checks in line 292 and 294 are too restrictive. One more concern is that users should call init<S64> to support S64 dtype, which looks verbose.

How about handling S32/S64 inside `extract_const`? It will simplify the function.

Reply:
Thank you very much for your detailed review!

> I think DType is not a good template parameter.  

I agree with your points. 

> How about handling S32/S64 inside extract_const? It will simplify the function.

Good idea. I updated this PR. Could you take another look?